### PR TITLE
Feature/2219 changelog

### PR DESCRIPTION
**Pull request by timohuber**

### CODE REVIEW

File: pool/app/changelog/entity.ml [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    match json with
+    | `Tuple [ a; b ] -> a, b
+    | _ -> raise (Exception "TODO")
+  ;;
+
+  let yojson_of_t (a, b) = `Tuple [ a; b ]
+end
+
+module Changes = struct
+  (* TODO: There has to be a better way *)
```
Comment (mabiede) [inline pool/app/changelog/entity.ml:22]: Add the link to the comment. Currently, the `@to_yojson/@of_joyson` is working for record types but not for variants.

[Documentation `@to_yojson/@of_joyson`](<https://github.com/ocaml-ppx/ppx_deriving_yojson?tab=readme-ov-file#to_yojson--of_yojson>)

File: pool/app/changelog/repo/repo_entity.ml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+module Field = struct
+  open Pool_message.Field
+
+  let t = make_type Caqti_type.string (read %> CCResult.return) show
+end
+
+let t =
+  let open Database.Caqti_encoders in
+  let encode _ = failwith "Read only model" in
```
Comment (mabiede) [inline pool/app/changelog/repo/repo_entity.ml:22]:
```suggestion
  let encode _ = 
    Pool_common.Utils.failwith Pool_message.Error.ReadOnlyModel
  in
```

File: pool/app/changelog/repo/repo_entity.ml [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        Data.
+          [ m.Write.id
+          ; m.model
+          ; m.entity_uuid
+          ; m.user_uuid
+          ; m.changes
+          ; m.created_at
+          ]
+    in
+    let decode _ = failwith "Write only model" in
```
Comment (mabiede) [inline pool/app/changelog/repo/repo_entity.ml:60]:
```suggestion
    let decode _ =
      Pool_common.Utils.failwith Pool_message.Error.WriteOnlyModel
    in
```

File: pool/app/organisational_unit/organisational_unit.mli [hunk 4]
```diff
@@ -68,3 +70,7 @@ val default_query : Query.t
 val filterable_by : Query.Filter.human option
 val searchable_by : Query.Column.t list
 val sortable_by : Query.Column.t list
+
+module Changelog : sig
+  include Changelog.TSig with type record = t
+end
```
Comment (mabiede) [inline pool/app/organisational_unit/organisational_unit.mli:76]:
```suggestion
module Changelog : Changelog.TSig with type record = t
```

File: pool/app/pool_database/migrations/migration_202408081359.ml [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+      PRIMARY KEY (id),
+      CONSTRAINT unique_uuid UNIQUE KEY (uuid)
+      ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci
+    |sql}
+;;
+
+(* let add_fk_contraint_to_users = Database.Migration.Step.create ~label:"add fk
+   contraint to users" {sql| ALTER TABLE pool_change_log ADD CONSTRAINT
+   fk_pool_changelog_user_users FOREIGN KEY (user_uuid) REFERENCES
+   user_users(uuid) |sql} ;; *)
```
Comment (mabiede) [inline pool/app/pool_database/migrations/migration_202408081359.ml:23]: comment

File: pool/app/pool_location/entity.ml [hunk 6]
```diff
@@ -81,18 +83,29 @@ module Status = struct
   let init = Active
 end
 
+module Files = struct
+  type t = Mapping.file list [@@deriving show]
+
+  let t_of_yojson _ = failwith "TODO"
```
Comment (mabiede) [inline pool/app/pool_location/entity.ml:89]: todo

File: pool/app/pool_location/pool_location.mli [hunk 7]
```diff
@@ -398,3 +400,7 @@ val default_query : Query.t
 val filterable_by : Query.Filter.human option
 val searchable_by : Query.Column.t list
 val sortable_by : Query.Column.t list
+
+module Changelog : sig
+  include Changelog.TSig with type record = t
+end
```
Comment (mabiede) [inline pool/app/pool_location/pool_location.mli:406]:
```suggestion
module Changelog : Changelog.TSig with type record = t
```

File: pool/app/changelog/changelog.ml [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        check [] json_list
+      in
+      match (json_before : Yojson.Safe.t), (json_after : Yojson.Safe.t) with
+      | `Assoc l1, `Assoc l2 ->
+        let keys =
+          let get = CCList.map fst in
+          get l1 @ get l2 |> CCList.uniq ~eq
+        in
+        keys
+        |> List.filter_map (fun key ->
```
[the remark this answers is not in the document]
Comment (timohuber) [reply pool/app/changelog/changelog.ml:49]: Opening `CCList` causes naming conflicts

File: pool/app/changelog/changelog.ml [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+  ;;
+
+  let make
+    ?(id = Id.create ())
+    ~entity_uuid
+    ~user_uuid
+    (before : R.t)
+    (after : R.t)
+    =
+    (*** the entity_uuid could also be part of the Reord module, as a function.
```
[the remark this answers is not in the document]
Comment (timohuber) [reply pool/app/changelog/changelog.ml:84]: That is still a WIP note 😄  
Still not sure if it is more convenient to use the `Entity` module (https://github.com/uzh/z-pool-tool/blob/feature/2219-changelog/pool/app/pool_location/version_history.ml) or to create e separate `Record` module.

File: pool/app/pool_location/entity.ml (hunk 6 above)
Comment (timohuber) [reply pool/app/pool_location/entity.ml:89]: added constraint

**Pull request closed without merging**